README
NAME
    Lego::Part::Image - Lego part image abstract class.

SYNOPSIS
     use Lego::Part::Image;

     my $obj = Lego::Part::Image->new;
     $obj->image;
     $obj->image_url;

METHODS
  * "new"
     my $obj = Lego::Part::Image->new;

    Constructor.

    *       "part"

            Lego::Part object.

            It is required.

            Default value is undef.

    Returns instance of object.

  "image"
     $obj->image;

    Abstract method for getting image.

  "image_url"
     $obj->image_url;

    Abstract method for getting image url.

ERRORS
     new():
             Parameter 'part' is required.
             Parameter 'part' must be Lego::Part object.
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

     image():
             This is abstract class. image() method not implemented.

     image_url():
             This is abstract class. image_url() method not implemented.

EXAMPLE
     use strict;
     use warnings;

     use Lego::Part;
     use Lego::Part::Image;

     # Error pure setting.
     $ENV{'ERROR_PURE_TYPE'} = 'Print';

     # Object.
     my $obj = Lego::Part::Image->new(
             'part' => Lego::Part->new(
                    'color' => 'red',
                    'design_id' => '3002',
             ),
     );

     # Get image.
     $obj->image;

     # Output:
     # Lego::Part::Image: This is abstract class. image() method not implemented.

DEPENDENCIES
    Class::Utils, Error::Pure, Readonly, Scalar::Util.

SEE ALSO
    Lego::Part::Image::BricklinkCom
        Lego part image class for bricklink.com.

    Lego::Part::Image::LegoCom
        Lego part image class for lego.com.

    Lego::Part::Image::LugnetCom
        Lego part image class for lugnet.com.

    Lego::Part::Image::PeeronCom
        Lego part image class for peeron.com.

    Task::Lego
        Install the Lego modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/Lego-Part-Image>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2015-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.07



Changes
0.07
 - Add error test for BricklinkCom.
 - Update SEE ALSO section in doc.

0.06 2023-11-25T11:24:40+01:00
 - Add support for bricklink.com.
 - Fix LegoCom URL.
 - Fix use lib '.' in Makefile.PL.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Move bugtracker to github.
 - Rename example files to better names.
 - Rewrite to use of =head2 in doc.
 - Update Module::Install to 1.21 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.05 2015-09-24T00:21:23+02:00
 - Add tests for image_url() methods.
 - Move author tests to xt/ directory.

0.04 2015-08-25T23:05:17+02:00
 - Fix Makefile.PL and META.yml files.
 - Update examples/ directory.

0.03 2015-08-24T11:28:31+02:00
 - Add Task::Lego to SEE ALSO section in doc.
 - Add minimal version of Perl.
 - Improve SEE ALSO sections in doc.
 - Update Module::Install to 1.16 version.

0.02 2015-03-09T06:18:53+01:00
 - Add tests for constructors.
 - Fix test dependencies.

0.01 2015-03-08T18:53:37+01:00
 - First version.


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Lego part image objects.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Lego-Part-Image';
readme_from 'Image.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.03;
requires 'Error::Pure' => 0.22;
requires 'Readonly' => 0;
requires 'Scalar::Util' => 0;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Lego-Part-Image/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Lego-Part-Image';
resources 'repository' => 'git://github.com/michal-josef-spacek/Lego-Part-Image';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'Lego::Part' => 0.01;
test_requires 'Test::MockObject' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.07';

# Run.
WriteAll();


Image/LegoCom.pm
package Lego::Part::Image::LegoCom;

use base qw(Lego::Part::Image);
use strict;
use warnings;

use Error::Pure qw(err);

our $VERSION = 0.07;

# Get image URL.
sub image_url {
	my $self = shift;

	if (! defined $self->{'part'}->element_id) {
		err "Element ID doesn't defined.";
	}
	my $url = sprintf 'https://www.lego.com/cdn/product-assets/element.img.lod5photo.192x192/%s.jpg',
		$self->{'part'}->element_id;

	return $url;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Lego::Part::Image::LegoCom - Lego part image class for lego.com.

=head1 SYNOPSIS

 use Lego::Part::Image::LegoCom;

 my $obj = Lego::Part::Image::LegoCom->new;
 my $image = $obj->image;
 my $image_url = $obj->image_url;

=head1 METHODS

=head2 C<new>

 my $obj = Lego::Part::Image::LegoCom->new;

Constructor.

=over 8

=item * C<part>

L<Lego::Part> object.

It is required.

Default value is undef.

=back

Returns instance of object.

=head2 C<image>

 my $image = $obj->image;

Get image.

Not implemented now.

=head2 C<image_url>

 my $image_url = $obj->image_url;

Get image URL.

Returns string with image URL.

=head1 ERRORS

 new():
         From Lego::Part::Image::new():
                 Parameter 'part' is required.
                 Parameter 'part' must be Lego::Part object.
                 From Class::Utils::set_params():
                         Unknown parameter '%s'.

 image():
         This is abstract class. image() method not implemented.

 image_url():
         Element ID doesn't defined.

=head1 EXAMPLE

=for comment filename=link_for_lego_com.pl

 use strict;
 use warnings;

 use Lego::Part;
 use Lego::Part::Image::LegoCom;

 # Object.
 my $obj = Lego::Part::Image::LegoCom->new(
         'part' => Lego::Part->new(
                'element_id' => '300321',
         ),
 );

 # Get image URL.
 my $image_url = $obj->image_url;

 # Print out.
 print "Part with element ID '300321' URL is: ".$image_url."\n";

 # Output:
 # Part with element ID '300321' URL is: https://www.lego.com/cdn/product-assets/element.img.lod5photo.192x192/300321.jpg

=begin html

<img src="https://www.lego.com/cdn/product-assets/element.img.lod5photo.192x192/300321.jpg" alt="Lego brick with element ID '300321'." />

=end html

=head1 DEPENDENCIES

L<Error::Pure>,
L<Lego::Part::Image>.

=head1 SEE ALSO

=over

=item L<Lego::Part::Image>

Lego part image abstract class.

=item L<Lego::Part::Image::PeeronCom>

Lego part image class for peeron.com.

=item L<Lego::Part::Image::LugnetCom>

Lego part image class for lugnet.com.

=item L<Task::Lego>

Install the Lego modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Lego-Part-Image>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2015-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut


Image/LugnetCom.pm
package Lego::Part::Image::LugnetCom;

use base qw(Lego::Part::Image);
use strict;
use warnings;

use Error::Pure qw(err);

our $VERSION = 0.07;

# Get image URL.
sub image_url {
	my $self = shift;

	if (! defined $self->{'part'}->design_id) {
		err "Design ID doesn't defined.";
	}
	my $url = sprintf 'http://img.lugnet.com/ld/%s.gif',
		$self->{'part'}->design_id;

	return $url;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Lego::Part::Image::LugnetCom - Lego part image class for lugnet.com.

=head1 SYNOPSIS

 use Lego::Part::Image::LugnetCom;

 my $obj = Lego::Part::Image::LugnetCom->new;
 my $image = $obj->image;
 my $image_url = $obj->image_url;

=head1 METHODS

=head2 C<new>

 my $obj = Lego::Part::Image::LugnetCom->new;

Constructor.

=over 8

=item * C<part>

L<Lego::Part> object.

It is required.

Default value is undef.

=back

Returns instance of object.

=head2 C<image>

 my $image = $obj->image;

Get image.

Not implemented now.

=head2 C<image_url>

 my $image_url = $obj->image_url;

Get image URL.

Returns string with image URL.

=head1 ERRORS

 new():
         From Lego::Part::Image::new():
                 Parameter 'part' is required.
                 Parameter 'part' must be Lego::Part object.
                 From Class::Utils::set_params():
                         Unknown parameter '%s'.

 image():
         This is abstract class. image() method not implemented.

 image_url():
         Design ID doesn't defined.

=head1 EXAMPLE

=for comment filename=link_for_lugnet_com.pl

 use strict;
 use warnings;

 use Lego::Part;
 use Lego::Part::Image::LugnetCom;

 # Object.
 my $obj = Lego::Part::Image::LugnetCom->new(
         'part' => Lego::Part->new(
                'design_id' => '3003',
         ),
 );

 # Get image URL.
 my $image_url = $obj->image_url;

 # Print out.
 print "Part with design ID '3003' URL is: ".$image_url."\n";

 # Output:
 # Part with design ID '3003' URL is: http://img.lugnet.com/ld/3003.gif

=begin html

<img src="http://img.lugnet.com/ld/3003.gif" alt="Lego brick with design ID '3003'." />

=end html

=head1 DEPENDENCIES

L<Error::Pure>,
L<Lego::Part::Image>.

=head1 SEE ALSO

=over

=item L<Lego::Part::Image>

Lego part image abstract class.

=item L<Lego::Part::Image::PeeronCom>

Lego part image class for peeron.com.

=item L<Lego::Part::Image::LegoCom>

Lego part image class for lego.com.

=item L<Task::Lego>

Install the Lego modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Lego-Part-Image>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2015-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut


Image/PeeronCom.pm
package Lego::Part::Image::PeeronCom;

use base qw(Lego::Part::Image);
use strict;
use warnings;

use Error::Pure qw(err);

our $VERSION = 0.07;

# Get image URL.
sub image_url {
	my $self = shift;

	if (! defined $self->{'part'}->color) {
		err "Color doesn't defined.";
	}
	if (! defined $self->{'part'}->design_id) {
		err "Design ID doesn't defined.";
	}
	my $url = sprintf 'http://media.peeron.com/ldraw/images/%d/100/%s.png',
		$self->{'part'}->color, $self->{'part'}->design_id;

	return $url;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Lego::Part::Image::PeeronCom - Lego part image class for peeron.com.

=head1 SYNOPSIS

 use Lego::Part::Image::PeeronCom;

 my $obj = Lego::Part::Image::PeeronCom->new;
 my $image = $obj->image;
 my $image_url = $obj->image_url;

=head1 METHODS

=head2 C<new>

 my $obj = Lego::Part::Image::PeeronCom->new;

Constructor.

=over 8

=item * C<part>

L<Lego::Part> object.

It is required.

Default value is undef.

=back

Returns instance of object.

=head2 C<image>

 my $image = $obj->image;

Get image.

Not implemented now.

=head2 C<image_url>

 my $image_url = $obj->image_url;

Get image URL.

Returns string with image URL.

=head1 ERRORS

 new():
         From Lego::Part::Image::new():
                 Parameter 'part' is required.
                 Parameter 'part' must be Lego::Part object.
                 From Class::Utils::set_params():
                         Unknown parameter '%s'.

 image():
         This is abstract class. image() method not implemented.

 image_url():
         Color doesn't defined.
         Design ID doesn't defined.

=head1 EXAMPLE

=for comment filename=link_for_peeron_com.pl

 use strict;
 use warnings;

 use Lego::Part;
 use Lego::Part::Image::PeeronCom;

 # Object.
 my $obj = Lego::Part::Image::PeeronCom->new(
         'part' => Lego::Part->new(
                'color' => 1,
                'design_id' => '3003',
         ),
 );

 # Get image URL.
 my $image_url = $obj->image_url;

 # Print out.
 print "Part with design ID '3003' and color '1' URL is: ".$image_url."\n";

 # Output:
 # Part with design ID '3003' and color '1' URL is: http://media.peeron.com/ldraw/images/1/100/3003.png

=begin html

<img src="http://media.peeron.com/ldraw/images/1/100/3003.png" alt="Lego brick with design ID '3003' and color '1'." />

=end html

=head1 DEPENDENCIES

L<Error::Pure>,
L<Lego::Part::Image>.

=head1 SEE ALSO

=over

=item L<Lego::Part::Image>

Lego part image abstract class.

=item L<Lego::Part::Image::LegoCom>

Lego part image class for lego.com.

=item L<Lego::Part::Image::LugnetCom>

Lego part image class for lugnet.com.

=item L<Task::Lego>

Install the Lego modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Lego-Part-Image>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2015-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut


Image/BricklinkCom.pm
package Lego::Part::Image::BricklinkCom;

use base qw(Lego::Part::Image);
use strict;
use warnings;

use Error::Pure qw(err);

our $VERSION = 0.07;

# Get image URL.
sub image_url {
	my $self = shift;

	if (! defined $self->{'part'}->design_id) {
		err "Design ID doesn't defined.";
	}
	my $url;
	if (defined $self->{'part'}->color) {
		$url = sprintf 'https://img.bricklink.com/ItemImage/PN/%s/%s.png',
			$self->{'part'}->color, $self->{'part'}->design_id;
	} else {
		$url = sprintf 'https://img.bricklink.com/ItemImage/PL/%s.png',
			$self->{'part'}->design_id;
	}

	return $url;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Lego::Part::Image::BricklinkCom - Lego part image class for bricklink.com.

=head1 SYNOPSIS

 use Lego::Part::Image::BricklinkCom;

 my $obj = Lego::Part::Image::BricklinkCom->new;
 my $image = $obj->image;
 my $image_url = $obj->image_url;

=head1 METHODS

=head2 C<new>

 my $obj = Lego::Part::Image::BricklinkCom->new;

Constructor.

=over 8

=item * C<part>

L<Lego::Part> object.

It is required.

Default value is undef.

=back

Returns instance of object.

=head2 C<image>

 my $image = $obj->image;

Get image.

Not implemented now.

=head2 C<image_url>

 my $image_url = $obj->image_url;

Get image URL.

Returns string with image URL.

=head1 ERRORS

 new():
         From Lego::Part::Image::new():
                 Parameter 'part' is required.
                 Parameter 'part' must be Lego::Part object.
                 From Class::Utils::set_params():
                         Unknown parameter '%s'.

 image():
         This is abstract class. image() method not implemented.

 image_url():
         Color doesn't defined.
         Design ID doesn't defined.

=head1 EXAMPLE

=for comment filename=link_for_bricklink_com.pl

 use strict;
 use warnings;

 use Lego::Part;
 use Lego::Part::Image::BricklinkCom;

 # Object.
 my $obj = Lego::Part::Image::BricklinkCom->new(
         'part' => Lego::Part->new(
                'color' => 1,
                'design_id' => '3003',
         ),
 );

 # Get image URL.
 my $image_url = $obj->image_url;

 # Print out.
 print "Part with design ID '3003' and color '1' URL is: ".$image_url."\n";

 # Output:
 # Part with design ID '3003' and color '1' URL is: https://img.bricklink.com/ItemImage/PN/1/3003.png

=begin html

<img src="https://img.bricklink.com/ItemImage/PN/1/3003.png" alt="Lego brick with design ID '3003' and color '1'." />

=end html

=head1 DEPENDENCIES

L<Error::Pure>,
L<Lego::Part::Image>.

=head1 SEE ALSO

=over

=item L<Lego::Part::Image>

Lego part image abstract class.

=item L<Lego::Part::Image::LegoCom>

Lego part image class for lego.com.

=item L<Lego::Part::Image::LugnetCom>

Lego part image class for lugnet.com.

=item L<Task::Lego>

Install the Lego modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Lego-Part-Image>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2015-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut


Image.pm
package Lego::Part::Image;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use Readonly;
use Scalar::Util qw(blessed);

# Constants.
Readonly::Scalar our $EMPTY_STR => q{};

our $VERSION = 0.07;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;	

	# Part object.
	$self->{'part'} = undef;

	# Process parameters.
	set_params($self, @params);

	# Check part object.
	if (! defined $self->{'part'}) {
		err "Parameter 'part' is required.";
	}
	if (! blessed($self->{'part'})
		|| ! $self->{'part'}->isa('Lego::Part')) {

		err "Parameter 'part' must be Lego::Part object.";
	}

	# Object.
	return $self;
}

# Get image.
sub image {
	my $self = shift;

	# TODO Implement getting of image with cache.
	err "This is abstract class. image() method not implemented.";

	return;
}

# Get image URL.
sub image_url {
	my $self = shift;

	err "This is abstract class. image_url() method not implemented.";

	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Lego::Part::Image - Lego part image abstract class.

=head1 SYNOPSIS

 use Lego::Part::Image;

 my $obj = Lego::Part::Image->new;
 $obj->image;
 $obj->image_url;

=head1 METHODS

=head2 * C<new>

 my $obj = Lego::Part::Image->new;

Constructor.

=over 8

=item * C<part>

L<Lego::Part> object.

It is required.

Default value is undef.

=back

Returns instance of object.

=head2 C<image>

 $obj->image;

Abstract method for getting image.

=head2 C<image_url>

 $obj->image_url;

Abstract method for getting image url.

=head1 ERRORS

 new():
         Parameter 'part' is required.
         Parameter 'part' must be Lego::Part object.
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

 image():
         This is abstract class. image() method not implemented.

 image_url():
         This is abstract class. image_url() method not implemented.

=head1 EXAMPLE

=for comment filename=abstract_class.pl

 use strict;
 use warnings;

 use Lego::Part;
 use Lego::Part::Image;

 # Error pure setting.
 $ENV{'ERROR_PURE_TYPE'} = 'Print';

 # Object.
 my $obj = Lego::Part::Image->new(
         'part' => Lego::Part->new(
                'color' => 'red',
                'design_id' => '3002',
         ),
 );

 # Get image.
 $obj->image;

 # Output:
 # Lego::Part::Image: This is abstract class. image() method not implemented.

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<Readonly>,
L<Scalar::Util>.

=head1 SEE ALSO

=over

=item L<Lego::Part::Image::BricklinkCom>

Lego part image class for bricklink.com.

=item L<Lego::Part::Image::LegoCom>

Lego part image class for lego.com.

=item L<Lego::Part::Image::LugnetCom>

Lego part image class for lugnet.com.

=item L<Lego::Part::Image::PeeronCom>

Lego part image class for peeron.com.

=item L<Task::Lego>

Install the Lego modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Lego-Part-Image>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2015-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut
